### ISSUE TITLE
Using a view as an enum table

### ISSUE THREAD

**Issue by ccakes** [graphile/postgraphile#1424]
### Summary

I _think_ there is an issue with using a view as an [enum table](https://www.graphile.org/postgraphile/enums/#with-enum-tables).

A test database with the example code from the documentation works fine but if I try the same thing using views and smart tags to signal the FK it's not recognised.

### Steps to reproduce

The `private` schema contains the raw data tables, the `public` schema is what we expose to Postgraphile.

```sql
CREATE DATABASE test;
\c test

CREATE SCHEMA private;
CREATE EXTENSION pgcrypto;

/**
 * private schema
 *
 * enum table
 */
CREATE TABLE private.status_values (
  id serial primary key,
  type text unique not null,
  description text
);

CREATE TABLE private.foo (
  id serial primary key,
  external_id uuid unique not null default gen_random_uuid(),
  name text unique not null,
  status int4 not null references private.status_values (id)
);

/**
 * public view
 */
CREATE VIEW public.status_values AS (
  SELECT type, description
  FROM private.status_values
);
COMMENT ON VIEW public.status_values IS E'@enum';

CREATE VIEW public.foo AS (
  SELECT
    f.external_id "id",
    f.name,
    s.type "status"
  FROM private.foo f
  JOIN private.status_values s ON s.id = f.status
);
COMMENT ON VIEW public.foo IS E'@foreignKey (status) references public.status_values (type)';
```

### Expected results

Queries and mutations for `public.foo` should use the enum type for the `status` field.

### Actual results

![image](https://user-images.githubusercontent.com/103831/103697142-cc3ee380-4f9f-11eb-8b05-67f4f4f8deaf.png)

### Additional context

- macOS Catalina
- PostgreSQL 13.1
- Node v14.15.0
- Postgraphile 4.10.0

### Possible Solution

No suggestions for a fix but if you have any suggestions for how to troubleshoot this further I'm happy to take a stab. I tried various combinations of `DEBUG` configs but couldn't find a way to see what the schema generator thinks that relationship looks like...

**Comment by ccakes** [comment]
```sql
CREATE TABLE public.counting (
  type text primary key,
  description text
);
COMMENT ON TABLE public.counting IS E'@enum';

INSERT INTO public.counting
VALUES
  ('One', '1'),
  ('Two', '2'),
  ('Three', '3');
```

It seems that during the schema build Postgraphile drops unused types - if I add ☝️ this to the schema above no enums appear but if I then add a table (not view) that uses it both appear as expected.

**Comment by benjie** [comment]
GraphQL schemas can only contain types that are used. If an enum is not used anywhere it can’t be in the schema.

Try adding a @primaryKey to your view.

**Comment by ccakes** [comment]
No change unfortunately. If I remove the `@enum` tag from `public.status_values` then the relationship is created fine so the `@foreignKey` tag is definitely working, seems to be limited to the fact that it's a view.

Oddly though it doesn't seem to matter which one is a view. If I change `public.status_values` to be a table (everything else the same) it doesn't work either.

When it's a table though, postgraphile throws an error on start up due to the enum table being empty. Obviously this makes sense, but putting some values in doesn't seem to make a difference. After that, I populated `private.status_values` with some data and tried again with all views but still no luck with the enum

**Comment by ccakes** [comment]
I did a bit of digging this morning and the issue appears to be that smart comment-based constraints aren't loaded at the time [enum relationships are introspected](https://github.com/graphile/graphile-engine/blob/0ef3e6bcab41a032a12748f4d73f5279b3359001/packages/graphile-build-pg/src/plugins/PgIntrospectionPlugin.js#L593).

At this point `result.constraint` only contains concrete constraints so theres nothing to link the enum table to. As an experiment I tried calling `smartCommentConstraints()` on `result` at line 573 and it seems to work... I assume that has other side effects but for my simple schema it seemed ok.

Just setting up a Graphile dev environment now to run the test suite.

**Comment by ccakes** [comment]
Late reply - but after that change the `postgraphile-core` test suite passes fine. Just running the entire suite now to be sure

**Comment by benjie** [comment]
Ah yeah; enum is the only "smart tag" that has to be done as a smart comment (as opposed to via the tags file); which means the other logics come later, and it cannot rely on other smart tags. I think this has to do with having to pull data in from the enum tables asynchronously, which can only be done as part of introspection (because everything after that has to be sync); but that doesn't mean we actually have to do the enum transform at that point I suppose...

**Comment by benjie** [comment]
Fixed in #704
